Text/CustomEntitySearch/CustomEntitySearch.cs

```diff
@@ -93,5 +95,146 @@ public static class CustomEntitySearch
             return new OkObjectResult(response);
         }
 
+        private static void AddValues(string checkMatch, string text, List<Entity> entities, HashSet<string> entitiesFound, int leniency)
+        {
+            bool addWord = false;
+            string escapedWord = Regex.Escape(checkMatch);
+            string pattern = @"\b(?ix:" + escapedWord + @")\b";
+            MatchCollection entityMatch = Regex.Matches(text, pattern, RegexOptions.IgnoreCase, TimeSpan.FromSeconds(MaxRegexEvalTime));
+            if (entityMatch.Count != 0)
+            {
+                foreach (Match match in entityMatch)
+                {
+                    entities.Add(
+                        new Entity
+                        {
+                            Category = "customEntity",
+                            Value = match.Value,
+                            Offset = match.Index,
+                            Confidence = 1
+                        });
+                }
+                addWord = true;
+            }
+
+            if (leniency > 0)
+            {
+                // using KMP for now
+                // Create Table
+                IList<int> KMPTable = new List<int>(new int[checkMatch.Length + 1]);
+                IList<char> wordCharArray = checkMatch.ToCharArray();
```

Comment:
imo this is algorithmically complicated enough to warrant its own unit tests. This is easiest to do when you refactor out the algorithmically significant code to its own class/helpers.